### PR TITLE
Sequential UUIDs and db optimizations.

### PR DESCRIPTION
**Pull request by onukristo**
## Context

Constant improvements are good.

### Changes

- We are starting to use sequential UUIDs, which are more suitable for database storage.
Gains are especially large and exponential on MariaDb.
1mln tasks 2x speed on db perf test.
2mln tasks 4x speed on db perf test.

Technically we use 38 bit timestamp (millis) prefix on random UUID as implicit task ids.

https://www.informit.com/articles/article.aspx?p=25862
https://www.2ndquadrant.com/en/blog/sequential-uuid-generators/
https://en.wikipedia.org/wiki/Universally_unique_identifier#As_database_keys

- (id,version) index was removed on Postgres as well, making db perf test to run 25% faster. 

- MariaDb schema for new services was redesigned.
However, the code is still working and keeps working with older schema as well.

- Another, more optimal table schema was tested and proposed for MariaDb applications which for whatever reasons are forced
to use random UUIDs with large number of tw tasks. 

- Added a db perf test to `demoapp` and `DemoAppRealTest`, which is more suitable to compare database bottlenecks tests.

- When a task is being set to a final state, the next_event_time is set to current time.
This will make the task cleaning process more accurate. 

## Checklist
- [ ] I have considered the impact of this change and added a [Change Classification](
https://transferwise.atlassian.net/wiki/spaces/EKB/pages/1401189673/Change+Classifications+and+Expectations) Label (`change:standard`, `change:impactful`, `change:emergency`)
- [ ] Change meets or does not compromise the [Baseline Security Requirements](https://transferwise.atlassian.net/wiki/spaces/EKB/pages/434929973/Baseline+Security+Requirements)

### CODE REVIEW

File: tw-tasks-core/src/main/java/com/transferwise/tasks/handler/TaskHandlerRegistry.java [hunk 1]
```diff
@@ -18,19 +22,29 @@
...
-    if (results.size() > 1) {
-      String resultsSt = StringUtils.join(results.stream().map(p -> p.getClass().getSimpleName()).collect(Collectors.toList()), ",");
-      throw new IllegalStateException("Too many handlers are able to handle task of type '" + task.getType() + "': " + resultsSt);
-    } else if (results.size() == 1) {
-      return results.get(0);
+    // TODO: Should we add caching here for larger applications?
+    // TODO: Or at least advice about keeping the number of handlers low in docs.
+    // TODO: Applications themselves can create hierarchical handlers themselves being able to handle most types of tasks. 
+    ITaskHandler result = null;
+    for (ITaskHandler taskHandler : getHandlers()) {
```
Comment (onukristo) [inline tw-tasks-core/src/main/java/com/transferwise/tasks/handler/TaskHandlerRegistry.java:36]: Profiler showed, that streams are too expensive...

File: tw-tasks-core/src/main/java/com/transferwise/tasks/utils/TwTasksUuidUtils.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import java.math.BigInteger;
+import java.util.UUID;
+import lombok.experimental.UtilityClass;
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.commons.lang3.StringUtils;
+
+@UtilityClass
+public class TwTasksUuidUtils {
+
+  public static UUID toUuid(String st) {
```
Comment (yevhenii0) [inline tw-tasks-core/src/main/java/com/transferwise/tasks/utils/TwTasksUuidUtils.java:13]: This is in purpose to support UUIDs that are not supported by `UUID.formString` or maybe `UUID.fromString` isn't efficient enough?

File: tw-tasks-core/src/main/java/com/transferwise/tasks/handler/TaskHandlerRegistry.java [hunk 3]
```diff
@@ -18,19 +22,29 @@
...
+      if (taskHandler.handles(task)) {
+        if (result != null) {
+          List<ITaskHandler> results = getHandlers().stream().filter(h -> h.handles(task)).collect(Collectors.toList());
+          String resultsSt = StringUtils.join(results.stream().map(p -> p.getClass().getSimpleName()).collect(Collectors.toList()), ",");
```
Comment (yevhenii0) [inline tw-tasks-core/src/main/java/com/transferwise/tasks/handler/TaskHandlerRegistry.java:40]: This one and above can be a single stream ending with `Collectors.joinining(",")`

File: tw-tasks-core/src/main/java/com/transferwise/tasks/handler/TaskHandlerRegistry.java [hunk 4]
```diff
@@ -18,19 +22,29 @@
...
   @Override
   public ITaskHandler getTaskHandler(IBaseTask task) {
-    List<ITaskHandler> results = getHandlers().stream().filter(h -> h.handles(task)).collect(Collectors.toList());
-
-    if (results.size() > 1) {
-      String resultsSt = StringUtils.join(results.stream().map(p -> p.getClass().getSimpleName()).collect(Collectors.toList()), ",");
-      throw new IllegalStateException("Too many handlers are able to handle task of type '" + task.getType() + "': " + resultsSt);
-    } else if (results.size() == 1) {
-      return results.get(0);
+    // TODO: Should we add caching here for larger applications?
```
Comment (yevhenii0) [inline tw-tasks-core/src/main/java/com/transferwise/tasks/handler/TaskHandlerRegistry.java:32]: Theoretically the application could have some state dependent implementations, probably not the case though.

**Review by yevhenii0 — APPROVED** [review]
What a boost!

File: tw-tasks-core/src/main/java/com/transferwise/tasks/utils/TwTasksUuidUtils.java (hunk 2 above)
Comment (onukristo) [reply tw-tasks-core/src/main/java/com/transferwise/tasks/utils/TwTasksUuidUtils.java:13]: UUID.fromString is too strict, yep.

File: tw-tasks-core/src/main/java/com/transferwise/tasks/handler/TaskHandlerRegistry.java (hunk 3 above)
Comment (onukristo) [reply tw-tasks-core/src/main/java/com/transferwise/tasks/handler/TaskHandlerRegistry.java]: yep, good idea. Will fix.

**Pull request merged**